### PR TITLE
[improve][functions] Check whether the resource exists and create it

### PR DESCRIPTION
**Pull request by nodece**
Signed-off-by: Zixuan Liu <nodeces@gmail.com>

### Motivation

Avoid printing the resource exists exception, the log like:
```
2022-06-28T15:36:13,509+0800 [broker-topic-workers-OrderedExecutor-2-0] ERROR org.apache.pulsar.broker.admin.v2.PersistentTopics - [null] Failed to create non-partitioned topic persistent://public/functions/coordinate
java.util.concurrent.CompletionException: org.apache.pulsar.broker.web.RestException: This topic already exists
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1159) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147) ~[?:?]
        at org.apache.pulsar.broker.service.BrokerService$2.lambda$openLedgerComplete$4(BrokerService.java:1426) ~[pulsar-broker.jar:2.11.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:795) ~[?:?]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: org.apache.pulsar.broker.web.RestException: This topic already exists
        at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.lambda$internalCreateNonPartitionedTopicAsync$36(PersistentTopicsBase.java:389) ~[pulsar-broker.jar:2.11.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150) ~[?:?]
        ... 12 more
```

### Modifications

- Add the tenant, cluster, and topic check

### Documentation

- [x] `doc-not-needed`

### CODE REVIEW

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/PulsarWorkerService.java [hunk 1]
```diff
@@ -389,7 +397,11 @@ private static Policies createFunctionsNamespacePolicies(String pulsarFunctionsC
...
         try {
-            getBrokerAdmin().topics().createNonPartitionedTopic(topic);
+            TopicName topicName = TopicName.get(topic);
+            List<String> topics = getBrokerAdmin().topics().getList(topicName.getNamespace());
+            if (topics.stream().noneMatch(n -> TopicName.get(n).equals(topicName))) {
+                getBrokerAdmin().topics().createNonPartitionedTopic(topic);
+            }
         } catch (PulsarAdminException e) {
             if (e instanceof PulsarAdminException.ConflictException) {
                 log.warn("Failed to create topic '{}': {}", topic, e.getMessage());
```
Comment (eolivelli) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/PulsarWorkerService.java:407]: this is the only line you want to change.
you can switch this to DEBUG

**Review by eolivelli — CHANGES_REQUESTED** [review]
I believe that this change does not fix your problem.

you can still have a race and have two workers that start and see the same problem.

I suggest to simply switch to "debug" level the line you are seeing (the error "topic already exists") as we are already doing for `AlreadyExistsException` for the tenants

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/PulsarWorkerService.java (hunk 1 above)
Comment (nodece) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/PulsarWorkerService.java:407]: Switch to the "debug" level is easier, but you can still see some logs in the broker.

**Comment by tisonkun** [comment]
Closed as stale and no consensus. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**